[LLDB][NFC] Added the interface DWARFExpression::Delegate to break dependencies and reduce lldb-server size

slydiman opened the pull request:
This patch addresses the issue #129543.

After this patch DWARFExpression does not call DWARFUnit directly and does not depend on lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp and a lot of clang code.

After this patch the size of lldb-server binary (Linux Aarch64) is reduced from 42MB to 13MB with LLVM 20.0.0
and from 47MB to 17MB with LLVM 21.0.0.

slydiman committed:
[LLDB][NFC] Added the interface DWARFUnitInterface to break dependencies and reduce lldb-server size

This patch addresses the issue #129543.

After this patch DWARFExpression does not call DWARFUnit directly and does not depend on lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp and a lot of clang code.

After this patch the size of lldb-server binary (Linux Aarch64)
is reduced from 42MB to 13MB with LLVM 20.0.0
and from 47MB to 17MB with LLVM 21.0.0.

labath reviewed (commented):
I think something like this makes sense, but if we're doing this, I think we should do it properly and actually remove the DWARFExpression->SymbolFileDWARF dependency. We can't do that if `DWARFUnitInterface` is defined in the symbol file plugin, nor while it contains methods like `GetSymbolFileDWARF()`. Moving the interface definition elsewhere is easy. The second part requires some thought. It looks like the main use for that method is to provide vendor extensibility for dwarf parsing, so I think we should expose *that* as an interface (i.e., have methods like ParseVendorDWARFOpcodeSize and GetVendorDWARFOpcodeSize) instead of the raw DWARF symbol file.

For this reason, I also think the interface shouldn't be called DWARFUnitInterface, but rather something centred on DWARFExpression and the things it needs from its users. For lack of imagination, I'm going to suggest DWARFExpression::Delegate (because we already use that term in some places), but maybe we'll come up with something more specific in the process. (This means it may not make sense to DWARFUnit to inherit from DWARFExpression::Delegate, but that's okay -- we can always use composition instead)

I also want to look at each function in that interface and see if it's really necessary. Two examples:
- GetLocationTable is only used from ParseDWARFLocationList and *that* is only used from DWARF plugin code. Instead of adding this to the interface, I think we should move ParseDWARFLocationList into the DWARF plugin. (This can/should be a separate patch)
- I'm not sure that `GetAddressByteSize` is necessary since the DataExtractor member (`m_data`) already has a method with the same name (and it's even used in some places). Before adding that to the interface, I'd like to see if we can make everything use the value in m_data instead.

slydiman committed:
Replaced DWARFUnitInterface with DWARFExpression::Delegate.
Moved ParseDWARFLocationList to DWARF plugin.

slydiman commented:
After this patch and #132274 the size of lldb-server is reduced from 47MB to 8MB.

labath commented on lldb/include/lldb/Expression/DWARFExpression.h:
```diff
@@ -40,6 +35,31 @@ class DWARFUnit;
+    virtual uint16_t GetVersion() const = 0;
+    virtual dw_addr_t GetBaseAddress() const = 0;
+    virtual uint8_t GetAddressByteSize() const = 0;
+    virtual llvm::Error GetDIEBitSizeAndSign(uint64_t die_offset,
+                                             uint64_t &bit_size,
+                                             bool &sign) = 0;
```
```suggestion
    virtual llvm::Expected<std::pair<uint64_t, bool>> GetDIEBitSizeAndSign(uint64_t relative_die_offset) const = 0;
```

(I'm prioritizing the interface over implementation, so I'm making the method const. You can put the const_cast into the implementation (I think it's only needed for `GetDIE`. I am also emphasizing the relativeness of the die offset as lldb default is to use absolute section offsets)

slydiman replied:
Done.

labath commented on lldb/source/Plugins/SymbolFile/DWARF/DWARFUnit.cpp:
```diff
@@ -696,6 +696,87 @@ llvm::StringRef DWARFUnit::PeekDIEName(dw_offset_t die_offset) {
+  if (!bit_size)
+    bit_size = die.GetAttributeValueAsUnsigned(DW_AT_bit_size, 0);
+  if (!bit_size)
+    return llvm::createStringError("unsupported type size in DW_OP_convert");
```
```suggestion
    return llvm::createStringError("unsupported type size");
```

Let's not mention DW_OP_convert here. (same below)

labath commented on lldb/source/Plugins/SymbolFile/DWARF/DWARFUnit.cpp:
```diff
@@ -696,6 +696,87 @@ llvm::StringRef DWARFUnit::PeekDIEName(dw_offset_t die_offset) {
+                                                     stack);
+}
+
+bool DWARFUnit::ParseDWARFLocationList(
+    const DataExtractor &data, DWARFExpressionList *location_list) const {
+  location_list->Clear();
```
Let's change this to a reference while we're here.

```suggestion
bool DWARFUnit::ParseDWARFLocationList(
    const DataExtractor &data, DWARFExpressionList &location_list) const {
  location_list.Clear();
```

labath commented on lldb/source/Plugins/SymbolFile/DWARF/DWARFUnit.h:
```diff
@@ -38,7 +39,7 @@ enum DWARFProducer {
 };
 
-class DWARFUnit : public UserID {
+class DWARFUnit : public UserID, public DWARFExpression::Delegate {
```
```suggestion
class DWARFUnit : public DWARFExpression::Delegate, public UserID {
```

(because Delegate is a "true" base class (with virtual methods and whatnot) while UserID is a mixin)

slydiman replied:
Done.

labath reviewed (commented):
I actually very much like this result. It'd be nice to (try) get rid of the `GetAddressByteSize` method, as I outlined before, but that's better done separately. Just a couple of small changes/improvements inline.

slydiman commented on lldb/source/Expression/DWARFExpression.cpp:
```diff
@@ -37,7 +36,6 @@
-#include "llvm/DebugInfo/DWARF/DWARFDebugLoc.h"
 #include "llvm/DebugInfo/DWARF/DWARFExpression.h"
 
 #include "Plugins/SymbolFile/DWARF/DWARFUnit.h"
```
Yes, because the code which is required this header has been moved to DWARFUnit.cpp

labath replied:
Could you remove it then? :)

slydiman committed:
Updated according to review.

slydiman committed:
Removed #include "Plugins/SymbolFile/DWARF/DWARFUnit.h" from DWARFExpression.cpp
Moved DW_OP_value_to_name() from DWARFDefines.cpp to DWARFExpression.cpp

slydiman commented:
@labath 
I have removed `#include "Plugins/SymbolFile/DWARF/DWARFUnit.h"` from DWARFExpression.cpp and moved the helper DW_OP_value_to_name() from DWARFDefines.cpp to DWARFExpression.cpp

felipepiovezan commented:
Should this also change the interface of DWAFExpressionList?

```
  DWARFExpressionList(lldb::ModuleSP module_sp,
                      const plugin::dwarf::DWARFUnit *dwarf_cu,
                      lldb::addr_t func_file_addr)
      : m_module_wp(module_sp), m_dwarf_cu(dwarf_cu),
        m_func_file_addr(func_file_addr) {}

  DWARFExpressionList(lldb::ModuleSP module_sp, DWARFExpression expr,
                      const plugin::dwarf::DWARFUnit *dwarf_cu)
      : m_module_wp(module_sp), m_dwarf_cu(dwarf_cu) {
    AddExpression(0, LLDB_INVALID_ADDRESS, expr);
  }
```

felipepiovezan commented:
This type of patch also slightly worries me in that it could make it harder to unify the LLDB dwarf data structures with the LLVM ones

slydiman commented:
> Should this also change the interface of DWAFExpressionList?

Yes, I will prepare the patch. Thanks

labath commented:
> This type of patch also slightly worries me in that it could make it harder to unify the LLDB dwarf data structures with the LLVM ones

That is something I considered, but I think this might actually help with that. At this point the LLDB and llvm dwarf classes are so different, that the only way I see to unify them is to do it piecemeal, and this kind of decoupling helps with that. So, if you wanted to replace LLDB DWARFUnit with llvm's, you could do that without touching DWARFExpression. It's true lldb's DWARFUnit now inherits from some DWARFExpression thingy, but that's easy to replace with composition. I alluded to that in the earlier comments, but in the end I didn't insist on it. Maybe I should have. If you think it's better to do that now, I'd be certainly fine with that.

This doesn't help with replacing LLDB's DWARFExpression with the llvm version, but I don't think it makes it worse either. I see a couple of possible ways forwards there:
- introduce a similar abstraction to llvm DWARFExpression. I think that would make sense since dwarf expressions are used in contexts (eh_frame) where there is no DWARFUnit to speak of.
- replace (non-unwinding) uses of DWARF Expression with a higher level abstraction. It's use in Function and Variable classes is breaking abstractions, and the PDB plugin has to jump through hoops to create DWARF expressions for these objects. That would leave DWARFExpression only used for unwinding in the core code, where it would be easier to replace this with llvm::DWARFExpression (and at that point, it may not matter that it depends on llvm::DWARFUnit, since this class -- unlike the lldb counterpart -- does not depend on the whole world)
